lighthouse-core/audits/time-to-interactive.js

```diff
@@ -0,0 +1,171 @@
+const SCORING_MEDIAN = 5000;
+
+class TTIMetric extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Performance',
+      name: 'time-to-interactive',
+      description: 'Time To Interactive',
+      optimalValue: SCORING_POINT_OF_DIMINISHING_RETURNS.toLocaleString(),
+      requiredArtifacts: ['traceContents', 'speedline']
+    };
+  }
+
+  /**
+   * Identify the time the page is "interactive"
+   * @see https://docs.google.com/document/d/1oiy0_ych1v2ADhyG_QW7Ps4BNER2ShlJjx2zCbVzVyY/edit#
+   *
+   * The user thinks the page is ready - (They believe the page is done enough to start interacting with)
+   *   - Layout has stabilized & key webfonts are visible.
+   *     AKA: First meaningful paint has fired.
+   *   - Page is nearly visually complete
+   *     Visual completion is 85%
+   *   - User-agent loading indicator is done
+   *     Current definition: Top frame and all iframes have fired window load event
+   *     Proposed definition (from cl/1860743002): top frame only: DCL ended and all layout-blocking resources, plus images that begun their request before DCL ended have finished.
+   *     Alternative definition (from Chrome on Android Progress Bar Enhancements - google-only, sry): top frame's DOMContentLoaded + top frame's images (who started before DCL) are loaded
+   *
+   * The page is actually ready for user:
+   *   - domContentLoadedEventEnd has fired
+   *     Definition: HTML parsing has finished, all DOMContentLoaded handlers have run.
+   *     No risk of DCL event handlers changing the page
+   *     No surprises of inactive buttons/actions as DOM element event handlers should be bound
+   *   - The main thread is available enough to handle user input
+   *     first 500ms window where Est Input Latency is <50ms at the 90% percentile.
+   * @param {!Artifacts} artifacts The artifacts from the gather phase.
+   * @return {!AuditResult} The score from the audit, ranging from 0-100.
+   */
+  static audit(artifacts) {
+    // We start looking at Math.Max(FMPMetric, visProgress[0.85])
+    return FMPMetric.audit(artifacts).then(fmpResult => {
+      if (fmpResult.value === -1) {
+        return generateError(fmpResult.debugString);
+      }
+      const fmpTiming = parseFloat(fmpResult.rawValue);
+      const timings = fmpResult.extendedInfo && fmpResult.extendedInfo.value &&
+          fmpResult.extendedInfo.value.timings;
+
+      // Process the trace
+      const tracingProcessor = new TracingProcessor();
+      const model = tracingProcessor.init(artifacts.traceContents);
+      const endOfTraceTime = model.bounds.max;
+
+      // TODO: Wait for DOMContentLoadedEndEvent
+      // TODO: Wait for UA loading indicator to be done
+
+      // TODO CHECK these units are the same
+      const fMPts = timings.fMPfull + timings.navStart;
+
+      // look at speedline results for 85% starting at FMP
+      const eightyFivePctVC = artifacts.Speedline.frames.find(frame => {
```

Comment:
looks like closure is unhappy here because there's no guarantee that a frame will be found that meets this conditional. A branch dealing with the `undefined` case (or a cast to `!SpeedlineFrames` if we're certain a frame will always exist meeting that criteria) should fix it